content/blog/2019-09-25-c-implementation.md

```diff
@@ -0,0 +1,189 @@
+it myself. It can be found in `src/table.{c,h}`. The dictionary uses open
+addressing and linear chaining. The hash function copies the hash used by
+java.lang.String’s hashCode() method. The `_hash` function computes
+a polynomial whos coeffecients are the integer values of the string's
+characters, evaluated at `x=31`. The polynomial is evaluated with Horner's
+method. The `table` maps string keys to `int64_t` values.  I represent bril
+integers and bril booleans with `int64_t` to avoid storing the type of bril
+variables. Since bril's typesystem only allows for those two types, this design
+is sufficient for now. Incorporating more types will be simple: I can extend
+the `table_elem` struct with a type bitfield/enum value.  For now, this simple
+table suffices. 
+
+The table code is also used for the label->index map and to store the mapping
+between string op codes and the index for that op code's implementing function
+in an array of function pointers.
+
+## Evaluation
+### Correctness
+I ran `cril` against the programs in `bril/test` and made sure the outputs
+matched `<program>.out`. My interpreter gave correct results on a fibonacci
+program posted in the class slack channel, on which the reference interpreter
+gave incorrect results. 
+
+### Performance
+I wanted to have a way to get reliable performance numbers to square off
+againts any other student's implementation of bril. First, I had the main
+evaluation loop return a `uint64_t` number of nanoseconds of elapsed time.
+I used the POSIX provided `timespec` structure to record the time tracked by
+`CLOCK_PROCESS_CPUTIME_ID`, the nanosecond resolution process time clock. This
+tracks CPU ticks spent on the program process itself, irrespective of other
+scheduled processes. The code is reproduced below:
+```c
+  JSON_Array *instrs = get_main_func_instrs(fn);
+  struct timespec tick, tock;
+  clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &tick);
+  mem = table_init();
+  labels = table_init();
+  make_dispatch();
+
+  num_instrs = json_array_get_count(instrs);
+  
+  for (size_t i = 0; i < num_instrs; ++i) {
+    curr_instr = json_array_get_object(instrs, i);
+    const char *label = json_object_get_string(curr_instr, "label");
+    if (label) {
+      table_put(labels, label, i);
+    }
+  }
+
+  size_t op_ind = -1;
+  char const *op = 0;
+  bool not_found = false;
+  for (ip = 0; ip < num_instrs; ++ip) {
+    curr_instr = json_array_get_object(instrs, ip);
+    op = jogs(curr_instr, "op");
+    if (!op) continue; // label
+    op_ind = (size_t) table_get(disp, op, &not_found);
+    if (not_found) {
+      quit("Operation not defined", op);
+    } else {
+      (op_funcs[op_ind]());
+    }
+  }
+
+  cleanup();
+  clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &tock);
+  return (1000000000 * (tock.tv_sec - tick.tv_sec) + tock.tv_nsec - tick.tv_nsec);
+```
+
+Notice the following:
+* I start timing on line 2, after JSON parsing. I did not want that to be
+  included
```

Comment:
That's fine, but keep in mind that, for a fair comparison, you'll also need to exclude JSON loading time in other implementations.